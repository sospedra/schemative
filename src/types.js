import { assign, reduce, noop } from 'lodash'
import PropTypes from 'prop-types'

const EXPOSED_PROPERTIES = ['value']

/**
 * Trigger a warning when assigning a value which does not
 * fit with the provided scalar.
 * Return default value.
 *
 * @param  {Any} candidate
 * @param  {Any} fallback
 * @param  {String} scalar
 * @return {Any}
 */
const evaluateValueType = (candidate, fallback, scalar) => {
  if (scalar !== 'any' && typeof candidate !== typeof fallback) {
    console.warn(
      'Evalute Schemative type expected',
      typeof fallback, 'but receive', typeof candidate,
      'when', candidate, 'value was set.',
      'Default value is returned instead.'
    )

    return fallback
  }

  return candidate
}

/**
 * Given a value, type base and default value creates an Schemative type.
 * Create the object with the type definition.
 * Pick all the properties that goes to the prototype.
 * Add the properties to an object containing the type exposed values.
 *
 * @param  {Any} value        - Selected value
 * @param  {Object} base      - Contains basic type's data
 *   @param  {String} scalar    - Type name
 *   @param  {Func} propTypes   - React PropTypes equivalent
 * @param  {Any} defaultValue - Default type value
 * @return {Object}           - Type
 */
const createType = (value, base, defaultValue) => {
  const type = assign({}, base, {
    isRequired: !base.__required__ &&
      createType(value, createTypeBase(base, true), defaultValue)
  })
  const properties = reduce(type, (memo, value, property) => {
    if (EXPOSED_PROPERTIES.includes(property)) return memo

    return assign({}, memo, {
      [property]: { get: () => value }
    })
  }, {})

  return Object.defineProperties({
    value: evaluateValueType(value, defaultValue, base.scalar)
  }, properties)
}

const createTypeBase = (base, isRequired) => {
  const propType = PropTypes[base.scalar]

  return {
    ...base,
    __required__: !!isRequired,
    propType: isRequired ? propType.isRequired : propType,
    scalar: base.scalar
  }
}

const createTypeWithDefault = (scalar, defaultValue) => {
  const base = createType(defaultValue, createTypeBase({ scalar }), defaultValue)
  const type = (value) => createType(value, base, defaultValue)

  Object.getOwnPropertyNames(base).forEach((property) => {
    if (EXPOSED_PROPERTIES.includes(property)) {
      type[property] = base[property]
    } else {
      Object.defineProperty(type, property, {
        get: () => base[property]
      })
    }
  })

  return type
}

export const any = createTypeWithDefault('any', true)
export const array = createTypeWithDefault('array', [])
export const bool = createTypeWithDefault('bool', true)
export const element = createTypeWithDefault('element', noop)
export const func = createTypeWithDefault('func', noop)
export const number = createTypeWithDefault('number', 0)
export const object = createTypeWithDefault('object', {})
export const shape = createTypeWithDefault('shape', {})
export const string = createTypeWithDefault('string', '')
export const symbol = createTypeWithDefault('symbol', Symbol(''))
