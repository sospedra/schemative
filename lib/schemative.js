'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createSchema = exports.shape = exports.oneOfType = exports.oneOf = exports.objectOf = exports.arrayOf = exports.symbol = exports.string = exports.object = exports.node = exports.number = exports.instanceOf = exports.func = exports.element = exports.bool = exports.array = exports.any = undefined;

var _react = require('react');

var _util = require('lodash/util');

var _util2 = _interopRequireDefault(_util);

var _parsers = require('./parsers');

var _transform = require('./transform');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var boundIsRequire = function boundIsRequire(prototype) {
  return Object.defineProperty(prototype, 'isRequired', {
    value: Object.assign({}, prototype, {
      type: prototype.type.isRequired
    })
  });
};

var createType = function createType(type, def) {
  return boundIsRequire({
    type: _react.PropTypes[type],
    default: def,
    iterable: false
  });
};

var createFunctionType = function createFunctionType(type) {
  return function (values) {
    return boundIsRequire({
      type: _react.PropTypes[type],
      values: values,
      iterable: true
    });
  };
};

var any = exports.any = createType('any', _util2.default.stubTrue());
var array = exports.array = createType('array', _util2.default.stubArray());
var bool = exports.bool = createType('bool', _util2.default.stubTrue());
var element = exports.element = createType('element', _util2.default.stubFalse());
var func = exports.func = createType('func', _util2.default.noop());
var instanceOf = exports.instanceOf = createType('instanceOf', _util2.default.stubFalse());
var number = exports.number = createType('number', -1);
var node = exports.node = createType('node', _util2.default.stubFalse());
var object = exports.object = createType('object', _util2.default.stubObject());
var string = exports.string = createType('string', _util2.default.stubString());
var symbol = exports.symbol = createType('symbol', Symbol);
var arrayOf = exports.arrayOf = createFunctionType('arrayOf');
var objectOf = exports.objectOf = createFunctionType('objectOf');
var oneOf = exports.oneOf = createFunctionType('oneOf');
var oneOfType = exports.oneOfType = createFunctionType('oneOfType');
var shape = exports.shape = createFunctionType('shape');

var createSchema = exports.createSchema = function createSchema(definition) {
  return {
    PropTypes: (0, _parsers.createPropTypes)(definition),
    Default: (0, _parsers.createDefault)(definition),
    transform: (0, _transform.createTransform)(definition)
  };
};
//# sourceMappingURL=schemative.js.map