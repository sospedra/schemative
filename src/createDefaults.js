import tsil from 'tsil'

export default (nodes) => {
  const defaultValues = tsil.modify(nodes, (node) => {
    return node.value
  })

  return tsil.deflatten(defaultValues)
}
