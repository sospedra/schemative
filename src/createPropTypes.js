import tsil from 'tsil'

export default (nodes) => {
  const propTypes = tsil.modify(nodes, (node) => {
    return node.propType
  })

  return tsil.deflatten(propTypes)
}
