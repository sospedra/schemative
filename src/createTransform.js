export default (nodes) => {
  // console.log('createTransform')
}
