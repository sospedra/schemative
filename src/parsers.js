import _ from 'lodash'

export const getPropType = (prop, attr) => {
  return prop.iterable ? recursiveValues(prop, attr) : prop[attr]
}

export const recursiveObject = (values, attr) => {
  return _.reduce(values, (memo, prop, key) => Object.assign({
    [key]: getPropType(prop, attr)
  }, memo), {})
}

export const recursiveArray = (values, attr) => {
  return _.map(values, (prop) => {
    return getPropType(prop, attr)
  })
}

export const selectRecursiveStrategy = (values, attr) => {
  return _.isArray(values)
    ? recursiveArray(values, attr)
    : recursiveObject(values, attr)
}

export const recursiveValues = ({ type, values }, attr) => {
  const exec = attr === 'type' ? type : _.identity
  return exec(selectRecursiveStrategy(values, attr))
}

export const createPropTypes = (definition) => {
  return recursiveObject(definition, 'type')
}

export const createDefault = (definition) => {
  return recursiveObject(definition, 'default')
}

export const filterByKeys = (definition, candidate) => {
  const keys = _.intersection(_.keys(definition), _.keys(candidate))
  return _.pick(candidate, keys)
}

export const transform = (definition, candidate, mutators) => Object.assign(
  filterByKeys(definition, candidate),
  _.reduce(mutators, (memo, value, key) => Object.assign({
    [key]: _.isFunction(value) ? value(candidate) : value,
    memo
  }), {})
)
