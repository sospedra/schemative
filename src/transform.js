import _ from 'lodash'

export const filterByKeys = (defsKeys, candidate) => {
  const keys = _.intersection(defsKeys, _.keys(candidate))
  return _.pick(candidate, keys)
}

export const createTransform = (definitions) => {
  const defsKeys = _.keys(definitions)
  let protoMutators = {}

  const transform = (candidate, mutators = protoMutators) => {
    return Object.assign(
      filterByKeys(defsKeys, candidate),
      _.reduce(mutators, (memo, value, key) => Object.assign({
        [key]: _.isFunction(value) ? value(candidate, definitions) : value
      }, memo), {})
    )
  }

  Object.defineProperty(transform, 'mutators', {
    get: () => protoMutators,
    set: (mutators) => {
      protoMutators = mutators
    }
  })

  return transform
}
