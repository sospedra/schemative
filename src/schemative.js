import { PropTypes } from 'react'
import _ from 'lodash/util'
import { createPropTypes, createDefault } from './parsers'
import { createTransform } from './transform'

const boundIsRequire = (prototype) => {
  return Object.defineProperty(prototype, 'isRequired', {
    value: Object.assign({}, prototype, {
      type: prototype.type.isRequired
    })
  })
}

const createType = (type, def) => boundIsRequire({
  type: PropTypes[type],
  default: def,
  iterable: false
})

const createFunctionType = (type) => (values) => boundIsRequire({
  type: PropTypes[type],
  values: values,
  iterable: true
})

export const any = createType('any', _.stubTrue())
export const array = createType('array', _.stubArray())
export const bool = createType('bool', _.stubTrue())
export const element = createType('element', _.stubFalse())
export const func = createType('func', _.noop())
export const instanceOf = createType('instanceOf', _.stubFalse())
export const number = createType('number', 0)
export const node = createType('node', _.stubFalse())
export const object = createType('object', _.stubObject())
export const string = createType('string', _.stubString())
export const symbol = createType('symbol', Symbol)
export const arrayOf = createFunctionType('arrayOf')
export const objectOf = createFunctionType('objectOf')
export const oneOf = createFunctionType('oneOf')
export const oneOfType = createFunctionType('oneOfType')
export const shape = createFunctionType('shape')

export const createSchema = (definition) => ({
  PropTypes: createPropTypes(definition),
  Default: createDefault(definition),
  transform: createTransform(definition)
})
