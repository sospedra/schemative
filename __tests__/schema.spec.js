/* global describe, it, expect */
import { PropTypes } from 'react'
import * as Schemative from '../src/schemative'
import * as mocks from './mocks'

describe('Suite types', () => {
  it('should have exist all the defined types', () => {
    mocks.types.forEach((type) => {
      expect(Schemative[type]).toMatchSnapshot()
    })

    mocks.functionTypes.forEach(([type]) => {
      expect(Schemative[type]).toMatchSnapshot()
    })
  })

  it('should be compatible with React.PropTypes', () => {
    mocks.types.forEach((type) => {
      expect(Schemative[type].type).toEqual(PropTypes[type])
    })

    mocks.functionTypes.forEach(([type, baseType]) => {
      expect(Schemative[type](baseType).type).toEqual(PropTypes[type])
    })
  })

  it('should allow isRequired extra prop', () => {
    mocks.types.forEach((type) => {
      expect(Schemative[type].isRequired).toMatchSnapshot()
      expect(Schemative[type].type.isRequired).toEqual(PropTypes[type].isRequired)
    })

    mocks.functionTypes.forEach(([type, baseType]) => {
      expect(Schemative[type].isRequired).toMatchSnapshot()
      expect(Schemative[type](baseType).type.isRequired).toEqual(PropTypes[type].isRequired)
    })
  })
})

describe('Suite createSchema', () => {
  it('should create from a simple non-nested definition', () => {
    expect(Schemative.createSchema(mocks.defs.simple)).toMatchSnapshot()
  })

  it('should create from a complex non-nested definition', () => {
    expect(Schemative.createSchema(mocks.defs.complex)).toMatchSnapshot()
  })

  it('should create from a simple nested definition', () => {
    expect(mocks.defs.simpleNested).toMatchSnapshot()
  })

  it('should create from a complex nested definition', () => {
    expect(mocks.defs.complexNested).toMatchSnapshot()
  })
})

describe('Suite schema', () => {
  const schema = Schemative.createSchema(mocks.defs.schemativePropTypes)

  it('should return the PropTypes', () => {
    expect(schema.PropTypes).toMatchSnapshot()
  })

  it('should return the default values', () => {
    expect(schema.Default).toMatchSnapshot()
  })

  it('should return a new fulfilled object after transform', () => {
    expect(schema.transform(mocks.fulfilled)).toMatchSnapshot()
  })

  it('should return a new fulfilled object with mutators', () => {
    expect(schema.transform(mocks.fulfilled, mocks.mutators)).toMatchSnapshot()
  })
})
