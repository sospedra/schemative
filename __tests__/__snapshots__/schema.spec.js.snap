exports[`Suite createSchema should create from a complex nested definition 1`] = `
Object {
  "geo": Object {
    "iterable": true,
    "type": [Function],
    "values": Object {
      "coord": Object {
        "default": -1,
        "iterable": false,
        "type": [Function],
      },
      "country": Object {
        "iterable": true,
        "type": [Function],
        "values": Object {
          "id": Object {
            "default": -1,
            "iterable": false,
            "type": [Function],
          },
        },
      },
      "flag": Object {
        "iterable": true,
        "type": [Function],
        "values": Array [
          Object {
            "default": "",
            "iterable": false,
            "type": [Function],
          },
          Object {
            "default": "",
            "iterable": false,
            "type": [Function],
          },
        ],
      },
    },
  },
}
`;

exports[`Suite createSchema should create from a complex non-nested definition 1`] = `
Object {
  "Default": Object {
    "contacts": Array [],
    "geo": Object {},
    "getName": undefined,
    "name": "",
    "year": -1,
  },
  "PropTypes": Object {
    "contacts": [Function],
    "geo": [Function],
    "getName": [Function],
    "name": [Function],
    "year": [Function],
  },
  "transform": [Function],
}
`;

exports[`Suite createSchema should create from a simple nested definition 1`] = `
Object {
  "contact": Object {
    "iterable": true,
    "type": [Function],
    "values": Array [
      Object {
        "default": -1,
        "iterable": false,
        "type": [Function],
      },
      Object {
        "default": -1,
        "iterable": false,
        "type": [Function],
      },
    ],
  },
  "name": Object {
    "default": "",
    "iterable": false,
    "type": [Function],
  },
}
`;

exports[`Suite createSchema should create from a simple non-nested definition 1`] = `
Object {
  "Default": Object {
    "name": "",
    "year": -1,
  },
  "PropTypes": Object {
    "name": [Function],
    "year": [Function],
  },
  "transform": [Function],
}
`;

exports[`Suite parsers should create an schema given an array 1`] = `Array []`;

exports[`Suite parsers should create an schema given an object 1`] = `Object {}`;

exports[`Suite parsers should create the schema default values 1`] = `
Object {
  "arr": Array [
    -1,
    "",
  ],
  "dispatch": undefined,
  "func": undefined,
  "shape": Object {
    "deep": Object {
      "nested": "",
    },
    "id": -1,
    "object": Object {},
  },
  "value": -1,
}
`;

exports[`Suite parsers should create the schema prop types 1`] = `
Object {
  "arr": [Function],
  "dispatch": [Function],
  "func": [Function],
  "shape": [Function],
  "value": [Function],
}
`;

exports[`Suite parsers should create the schema prop types with isRequired 1`] = `
Object {
  "geo": [Function],
  "req": [Function],
}
`;

exports[`Suite parsers should select an strategy for recursivity 1`] = `Object {}`;

exports[`Suite parsers should select an strategy for recursivity 2`] = `Object {}`;

exports[`Suite parsers should transform a definition using mutators 1`] = `
Object {
  "memo": Object {
    "extra": "attribute",
    "memo": Object {},
  },
  "name": "Ada Lovelace",
  "year": 1852,
}
`;

exports[`Suite schema should return a new fulfilled object after transform 1`] = `Object {}`;

exports[`Suite schema should return a new fulfilled object with mutators 1`] = `
Object {
  "memo": Object {
    "extra": "attribute",
    "memo": Object {},
  },
  "year": 1852,
}
`;

exports[`Suite schema should return the PropTypes 1`] = `
Object {
  "arr": [Function],
  "dispatch": [Function],
  "func": [Function],
  "shape": [Function],
  "value": [Function],
}
`;

exports[`Suite schema should return the default values 1`] = `
Object {
  "arr": Array [
    -1,
    "",
  ],
  "dispatch": undefined,
  "func": undefined,
  "shape": Object {
    "deep": Object {
      "nested": "",
    },
    "id": -1,
    "object": Object {},
  },
  "value": -1,
}
`;

exports[`Suite types should allow isRequired extra prop 1`] = `
Object {
  "default": Array [],
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 2`] = `
Object {
  "default": true,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 3`] = `
Object {
  "default": undefined,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 4`] = `
Object {
  "default": -1,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 5`] = `
Object {
  "default": Object {},
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 6`] = `
Object {
  "default": "",
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 7`] = `
Object {
  "default": [Function],
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 8`] = `
Object {
  "default": true,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 9`] = `
Object {
  "default": false,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 10`] = `
Object {
  "default": false,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 11`] = `
Object {
  "default": false,
  "isRequired": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should allow isRequired extra prop 12`] = `undefined`;

exports[`Suite types should allow isRequired extra prop 13`] = `undefined`;

exports[`Suite types should allow isRequired extra prop 14`] = `undefined`;

exports[`Suite types should allow isRequired extra prop 15`] = `undefined`;

exports[`Suite types should allow isRequired extra prop 16`] = `undefined`;

exports[`Suite types should have exist all the defined types 1`] = `
Object {
  "default": Array [],
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 2`] = `
Object {
  "default": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 3`] = `
Object {
  "default": undefined,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 4`] = `
Object {
  "default": -1,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 5`] = `
Object {
  "default": Object {},
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 6`] = `
Object {
  "default": "",
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 7`] = `
Object {
  "default": [Function],
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 8`] = `
Object {
  "default": true,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 9`] = `
Object {
  "default": false,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 10`] = `
Object {
  "default": false,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 11`] = `
Object {
  "default": false,
  "iterable": false,
  "type": [Function],
}
`;

exports[`Suite types should have exist all the defined types 12`] = `[Function]`;

exports[`Suite types should have exist all the defined types 13`] = `[Function]`;

exports[`Suite types should have exist all the defined types 14`] = `[Function]`;

exports[`Suite types should have exist all the defined types 15`] = `[Function]`;

exports[`Suite types should have exist all the defined types 16`] = `[Function]`;
